File: cpp/training/src/bloch_attention.cpp
#include "qminiwasm/training/bloch_attention.hpp"

namespace qminiwasm::training {
namespace {

torch::Tensor stokes_from_logits(torch::Tensor logits) {
  auto t = torch::tanh(logits);
  auto n = t.norm(2, /*dim=*/-1, /*keepdim=*/true).clamp_min(1e-6f);
  return t / n;
}

}  // namespace

BlochSphereAttentionImpl::BlochSphereAttentionImpl(std::int64_t d_model, int num_heads, std::int64_t d_value,
                                                   bool bias)
    : d_model_(d_model), num_heads_(num_heads) {
  TORCH_CHECK(num_heads > 0, "num_heads must be positive");
  if (d_value < 0) {
    TORCH_CHECK(d_model % num_heads == 0, "d_model must be divisible by num_heads when d_value is implicit");
    d_value_ = d_model / num_heads;
  } else {
    d_value_ = d_value;
  }
  q_proj_ = register_module("q_proj", torch::nn::Linear(
                                           torch::nn::LinearOptions(d_model_, num_heads_ * 3).bias(bias)));
  k_proj_ = register_module("k_proj", torch::nn::Linear(
                                           torch::nn::LinearOptions(d_model_, num_heads_ * 3).bias(bias)));
  v_proj_ = register_module(
      "v_proj", torch::nn::Linear(torch::nn::LinearOptions(d_model_, num_heads_ * d_value_).bias(bias)));
  out_proj_ = register_module(
      "out_proj", torch::nn::Linear(torch::nn::LinearOptions(num_heads_ * d_value_, d_model_).bias(bias)));
}

torch::Tensor BlochSphereAttentionImpl::forward(torch::Tensor x) {
  TORCH_CHECK(x.dim() == 3, "expected [B,T,D], got dim ", x.dim());
  const auto b = x.size(0);
  const auto t = x.size(1);
  const auto h = static_cast<std::int64_t>(num_heads_);
  auto q_logits = q_proj_->forward(x);
  auto k_logits = k_proj_->forward(x);
  auto q = stokes_from_logits(q_logits.view({b, t, h, 3}));
  auto k = stokes_from_logits(k_logits.view({b, t, h, 3}));
  auto v = v_proj_->forward(x).view({b, t, h, d_value_});
  // fidelity [B, T, S, H] then permute to [B, H, T, S]
  auto qe = q.unsqueeze(2);
  auto ke = k.unsqueeze(1);
  auto fid_ts = (1.0 + (qe * ke).sum(-1)) * 0.5;
  auto fid = fid_ts.permute({0, 3, 1, 2}).clamp_min(1e-6);
  auto w = fid / fid.sum(-1, true);
  auto vp = v.permute({0, 2, 1, 3});
  const auto dv = d_value_;
  auto w_flat = w.reshape({-1, t, t});
  auto v_flat = vp.reshape({-1, t, dv});
  auto ctx = torch::bmm(w_flat, v_flat);
  auto ctx_bthd = ctx.reshape({b, h, t, dv}).permute({0, 2, 1, 3}).reshape({b, t, h * dv});
  return out_proj_->forward(ctx_bthd);
}

}  // namespace qminiwasm::training

File: qminiwasm/layers/bloch_attention.py
"""Bloch-sphere fidelity attention (classical Mode A).

Maps per-head queries and keys to unit Stokes vectors on the Bloch sphere and uses
quantum fidelity ``F = (1 + s_q · s_k) / 2`` as nonnegative attention weights.
Replaces softmax with a normalized fidelity matrix over the sequence dimension.

QPU compute-uncompute execution (Mode B) is not wired here; see architecture docs.
"""

from __future__ import annotations

import math

import torch
import torch.nn as nn
import torch.nn.functional as F


def _stokes_from_logits(logits: torch.Tensor) -> torch.Tensor:
    """Map ``[..., 3]`` unconstrained logits to unit 3-vectors (Stokes / Bloch direction)."""
    return F.normalize(torch.tanh(logits), dim=-1, eps=1e-6)


class BlochSphereAttention(nn.Module):
    """Multi-head fidelity attention for sequences ``[batch, seq, d_model]``."""

    def __init__(
        self,
        d_model: int,
        num_heads: int,
        d_value: int | None = None,
        bias: bool = True,
        device: torch.device | None = None,
        dtype: torch.dtype | None = None,
    ) -> None:
        super().__init__()
        if num_heads <= 0:
            raise ValueError("num_heads must be positive")
        factory = {"device": device, "dtype": dtype}
        self.d_model = d_model
        self.num_heads = num_heads
        if d_value is None:
            if d_model % num_heads != 0:
                raise ValueError("d_model must be divisible by num_heads when d_value is None")
            d_value = d_model // num_heads
        self.d_value = d_value
        self.q_proj = nn.Linear(d_model, num_heads * 3, bias=bias, **factory)
        self.k_proj = nn.Linear(d_model, num_heads * 3, bias=bias, **factory)
        self.v_proj = nn.Linear(d_model, num_heads * d_value, bias=bias, **factory)
        self.out_proj = nn.Linear(num_heads * d_value, d_model, bias=bias, **factory)
        self.scale = 1.0 / math.sqrt(2.0)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        """Args:
            x: ``[batch, seq_len, d_model]``
        Returns:
            ``[batch, seq_len, d_model]``
        """
        if x.dim() != 3:
            raise ValueError(f"expected [B,T,D], got {tuple(x.shape)}")
        b, t, _d = x.shape
        h = self.num_heads
        q = _stokes_from_logits(self.q_proj(x).view(b, t, h, 3))
        k = _stokes_from_logits(self.k_proj(x).view(b, t, h, 3))
        v = self.v_proj(x).view(b, t, h, self.d_value)
        # fidelity [B, H, Tq, Tk]
        fid = (1.0 + torch.einsum("bthd,bshd->bhts", q, k)) * 0.5
        fid = torch.clamp(fid, min=1e-6)
        w = fid / fid.sum(dim=-1, keepdim=True)
        ctx = torch.einsum("bhts,bshd->bthd", w, v)
        ctx = ctx.reshape(b, t, h * self.d_value)
        return self.out_proj(ctx)
